nmtwizard/preprocess/operators/noise.py

```diff
@@ -0,0 +1,68 @@
+import random
+
+from nmtwizard.preprocess import prepoperator
+from nmtwizard.preprocess.tu import TokReplace
+
+@prepoperator.register_operator("noise")
+class Noise(prepoperator.TUOperator):
+
+    @staticmethod
+    def is_applied_for(process_type):
+        return process_type == prepoperator.ProcessType.TRAINING
+
+    def __init__(self, config, *args):
+        self._drop_word_prob = config.get("drop_word_prob", 0)
+        self._drop_space_prob = config.get("drop_space_prob", 0)
+        self._drop_char_prob = config.get("drop_char_prob", 0)
+        self._duplicate_char_prob = config.get("duplicate_char_prob", 0)
+        self._swap_char_prob = config.get("swap_char_prob", 0)
+
+    def _preprocess_tu(self, tu, *args):
+        tokens = tu.src_tok.token_objects
+
+        index_to_delete = self._apply_word_noise(tokens[0])
+        if index_to_delete:
+            self._delete_tokens(tu, index_to_delete)
+        else:
+            # Invalidate detokenized representation because some token surface may have changed.
+            tu.src_detok = None
```

Comment:
In my original design, the tokens returned by tok setter where actually a copy and couldn't be modified directly.
https://github.com/OpenNMT/nmt-wizard-docker/blob/27a799f50239302541c913eca4fb9eaa7e13cab6/nmtwizard/preprocess/tu.py#L102

The only way to modify the original tokens was via replace_tokens function that was supposed to automatically invalidate the detokenized version whenever the tokens where modified.

This was precisely to remove the need for each operator to think/implement this sync explicitly.

But maybe it's too much copying ?

Reply:
As discussed, we can avoid using `replace_tokens` in this module and simply reset the tokens.